sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -536,6 +538,134 @@ class CollationStringExpressionsSuite
+      StringTrimTestCase("UNICODE_CI", "TRIM", "asd", false, null, "asd"),
+      StringTrimTestCase("UNICODE_CI", "TRIM", "  asd  ", false, null, "asd"),
+      StringTrimTestCase("UNICODE_CI", "BTRIM", "asd", false, null, "asd"),
+      StringTrimTestCase("UNICODE_CI", "BTRIM", "  asd  ", false, null, "asd"),
+      StringTrimTestCase("UNICODE_CI", "LTRIM", "asd", false, null, "asd"),
+      StringTrimTestCase("UNICODE_CI", "LTRIM", "  asd  ", false, null, "asd  "),
+      StringTrimTestCase("UNICODE_CI", "RTRIM", "asd", false, null, "asd"),
+      StringTrimTestCase("UNICODE_CI", "RTRIM", "  asd  ", false, null, "  asd"),
+
+      // With null trimString param.
+      StringTrimTestCase("UNICODE_CI", "TRIM", "asd", true, null, null),
+      StringTrimTestCase("UNICODE_CI", "TRIM", "  asd  ", true, null, null),
+      StringTrimTestCase("UNICODE_CI", "BTRIM", "asd", true, null, null),
+      StringTrimTestCase("UNICODE_CI", "BTRIM", "  asd  ", true, null, null),
+      StringTrimTestCase("UNICODE_CI", "LTRIM", "asd", true, null, null),
+      StringTrimTestCase("UNICODE_CI", "LTRIM", "  asd  ", true, null, null),
+      StringTrimTestCase("UNICODE_CI", "RTRIM", "asd", true, null, null),
+      StringTrimTestCase("UNICODE_CI", "RTRIM", "  asd  ", true, null, null),
+
+      // With " " trimString param.
+      StringTrimTestCase("UNICODE_CI", "TRIM", "asd", true, " ", "asd"),
+      StringTrimTestCase("UNICODE_CI", "TRIM", "  asd  ", true, " ", "asd"),
+      StringTrimTestCase("UNICODE_CI", "BTRIM", "asd", true, " ", "asd"),
+      StringTrimTestCase("UNICODE_CI", "BTRIM", "  asd  ", true, " ", "asd"),
+      StringTrimTestCase("UNICODE_CI", "LTRIM", "asd", true, " ", "asd"),
+      StringTrimTestCase("UNICODE_CI", "LTRIM", "  asd  ", true, " ", "asd  "),
+      StringTrimTestCase("UNICODE_CI", "RTRIM", "asd", true, " ", "asd"),
+      StringTrimTestCase("UNICODE_CI", "RTRIM", "  asd  ", true, " ", "  asd")
+
+      // Other more complex cases can be found in unit tests in CollationSupportSuite.java.
+    )
+
+    // scalastyle:off
+
+    testCases.foreach(testCase => {
+      var df: DataFrame = null
+
+      if (testCase.trimFunc.equalsIgnoreCase("BTRIM")) {
+        // BTRIM has arguments in (srcStr, trimStr) order
+        df = sql(s"SELECT ${testCase.trimFunc}(" +
+          s"COLLATE('${testCase.sourceString}', '${testCase.collation}')" +
+          (if (!testCase.hasTrimString) "" else if (testCase.trimString == null) ", null" else s", COLLATE('${testCase.trimString}', '${testCase.collation}')") +
+          ")")
+      }
+      else {
+        // While other functions have arguments in (trimStr, srcStr) order
+        df = sql(s"SELECT ${testCase.trimFunc}(" +
+          (if (!testCase.hasTrimString) "" else if (testCase.trimString == null) "null, " else s"COLLATE('${testCase.trimString}', '${testCase.collation}'), ") +
+          s"COLLATE('${testCase.sourceString}', '${testCase.collation}')" +
+          ")")
+      }
+
```

Comment:
please take a look at some other tests in this suite (i.e. Contains/StartsWith/EndsWith)

if you made changes to collation type casts, then we need to test implicit casting here

Reply:
Yeah, I created initial set of tests now, but I'm not yet sure what exactly happened after Mihailo's changes regarding casting. I'll need to sync with him to understand better his changes before adding more related tests.

Earlier we had type mismatch checks in `stringExpressions`, but I see he has removed those, so I didn't understand that part yet. I'll sync with him and update the PR accordingly afterwards.

Reply:
when you added `_: StringTrim | _: StringTrimLeft | _: StringTrimRight` to that last case in CollationTypeCasts, that should ensure that casting is enforced

so now just add tests here to confirm that it works

Reply:
like just modify the query to do `'${testCase.sourceString}'` instead of `COLLATE('${testCase.sourceString}', '${testCase.collation}')"` and it should work the same if COLLATE is called on the other parameter

Reply:
Done.